tests/Microsoft.Identity.Test.Unit/CacheTests/CacheSyncronizationTests.cs

```diff
@@ -0,0 +1,123 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Identity.Client;
+using Microsoft.Identity.Test.Common.Core.Mocks;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace Microsoft.Identity.Test.Unit.CacheTests
+{
+    [TestClass]
+    public class CacheSyncronizationTests : TestBase
+    {
+
+        [TestMethod]
+        [Timeout(1000)]
+        public async Task DisableInternalSemaphore_Async()
+        {
+            await RunSemaphoreTestAsync(true).ConfigureAwait(false);
+            await RunSemaphoreTestAsync(false).ConfigureAwait(false);
+        }
+
+        private async Task RunSemaphoreTestAsync(bool useCacheSyncronization)
+        {
+            using (var harness = base.CreateTestHarness())
+            {
+                MockHttpManager httpManager = harness.HttpManager;
+                httpManager.AddInstanceDiscoveryMockHandler();
+
+                ConfidentialClientApplication app =
+                    ConfidentialClientApplicationBuilder.Create(TestConstants.ClientId)
+                                                              .WithClientSecret(TestConstants.ClientSecret)
+                                                              .WithAuthority(TestConstants.AuthorityUtidTenant)
+                                                              .WithCacheSyncronization(useCacheSyncronization)
+                                                              .WithHttpManager(httpManager)
+                                                              .BuildConcrete();
+
+                httpManager.AddMockHandlerSuccessfulClientCredentialTokenResponseMessage();
+
+                BlockingCache inMemoryTokenCache = new BlockingCache();
+                inMemoryTokenCache.Bind(app.AppTokenCache);
+
+                // Seed the cache with a token
+                var result = await app.AcquireTokenForClient(TestConstants.s_scope.ToArray())
+                    .ExecuteAsync()
+                    .ConfigureAwait(false);
+                Assert.IsTrue(result.AuthenticationResultMetadata.TokenSource == TokenSource.IdentityProvider);
+
+
+                var blockingTask = RunAsync(inMemoryTokenCache, app, true);
+                var nonBlockingTask1 = RunAsync(inMemoryTokenCache, app, false);
+                var nonBlockingTask2 = RunAsync(inMemoryTokenCache, app, false);
+
+                int res = Task.WaitAny(new[] { blockingTask, nonBlockingTask1, nonBlockingTask2 }, 100);
```

Comment:
CC @jmprieur for an technical explanation of the improvement here. In an app where CCA is singleton, a blocking cache operation would paralyze the entire system. But without the semaphore, some requests would go though.